data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/S3SelectObjectWorker.java

```diff
@@ -0,0 +1,236 @@
+    private final String queryStatement;
+    private final S3SelectSerializationFormatOption serializationFormatOption;
+    private final S3SelectResponseHandler s3SelectResponseHandler;
+    private final S3ObjectPluginMetrics s3ObjectPluginMetrics;
+    private final CompressionType compressionType;
+    private final FileHeaderInfo fileHeaderInfo;
+    private final BiConsumer<Event, S3ObjectReference> eventConsumer;
+    private final ObjectMapper mapper = new ObjectMapper();
+    private final JsonFactory jsonFactory = new JsonFactory();
+    public S3SelectObjectWorker(final S3ObjectRequest s3ObjectRequest) {
+        this.buffer = s3ObjectRequest.getBuffer();
+        this.numberOfRecordsToAccumulate = s3ObjectRequest.getNumberOfRecordsToAccumulate();
+        this.bufferTimeout = s3ObjectRequest.getBufferTimeout();
+        this.queryStatement = s3ObjectRequest.getQueryStatement();
+        this.serializationFormatOption = s3ObjectRequest.getSerializationFormatOption();
+        this.s3AsyncClient = s3ObjectRequest.getS3AsyncClient();
+        this.s3SelectResponseHandler = s3ObjectRequest.getS3SelectResponseHandler();
+        this.s3ObjectPluginMetrics = s3ObjectRequest.getS3ObjectPluginMetrics();
+        this.compressionType = s3ObjectRequest.getCompressionType();
+        this.fileHeaderInfo = s3ObjectRequest.getFileHeaderInfo();
+        this.eventConsumer = s3ObjectRequest.getEventConsumer();
+    }
+
+    /**
+     * Fetch the S3 object content using s3 select service and parsing and pushing to buffer.
+     * @param s3ObjectReference Contains bucket and s3 object details
+     *
+     * @throws IOException When Input/Output exception occur while processing the data.
+     */
+    public void parseS3Object(final S3ObjectReference s3ObjectReference) throws IOException {
+        try{
+            selectObjectContent(s3ObjectReference);
+        }catch(Exception e){
+            s3ObjectPluginMetrics.getS3ObjectsFailedCounter().increment();
+            throw new IOException(e);
+        }
+    }
+
+    private void selectObjectContent(S3ObjectReference s3ObjectReference) throws IOException {
+        final long totalBytesRead;
+        final InputStream inputStreamList;
+        final BufferAccumulator<Record<Event>> bufferAccumulator = BufferAccumulator.create(buffer, numberOfRecordsToAccumulate, bufferTimeout);
+        final InputSerialization inputSerialization = getInputSerializationFormat(serializationFormatOption
+                ,compressionType,fileHeaderInfo);
+        final SelectObjectContentRequest selectObjectContentRequest = getS3SelectObjRequest(s3ObjectReference, inputSerialization);
+        s3AsyncClient.selectObjectContent(selectObjectContentRequest, s3SelectResponseHandler).join();
+        final List<SelectObjectContentEventStream> receivedEvents = s3SelectResponseHandler.getReceivedEvents();
+        if (receivedEvents.isEmpty()) {
+            throw new IOException("No Records Found.");
```

Comment:
I think that throwing an exception here is not right. I'm not sure this is actually an exceptional issue. Because the user is providing a select statement, it is possible that some objects will return empty (say they use a [WHERE clause](https://docs.aws.amazon.com/AmazonS3/latest/userguide/s3-select-sql-reference-select.html#s3-glacier-select-sql-reference-where) which eliminates data.).

It may be reasonable to log an INFO log. Please include the `s3ObjectReference` in the log.
Something like:

```
log.info("S3 Select returned no events for S3 object {}, s3ObjectReference);
```

Reply:
Thanks David for your review, Changes are incorporated and available for review.

Reply:
Thanks David for your review, Changes are incorporated and available for review.